File: 서예진A/20230517/가장큰수.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

/* 
조합하면서 크기 비교하기 6+10 vs 10+6 => 610 > 106 , 610선택하도록
문자 결합해보고 사전 순서가 더 뒤인것(더 큰 것이 순서가 더 뒤) => 앞에 있도록 sort
return true => 순서를 유지한다
*/
bool compare(string &a, string &b) {
    return a+b > b+a;
}
string solution(vector<int> numbers) {
    string answer = "";
    
    //문자열로 바꾸어 return 하기 위해 vector<string>으로 만들기
    vector<string> v;
    
    for(int i=0;i<numbers.size();i++){
        v.push_back(to_string(numbers[i]));
    }
    sort(v.begin(), v.end(), compare); //compare함수에 기반하여 sort
    
    //원소가 0일 경우를 고려해야함
    
    for(int i=0; i<numbers.size();i++){
        answer+=v[i];
    }
    if(answer[0]=='0')
        answer='0'; //ex) 000일 경우 그냥 0으로 처리할 수 있도록 함
    
    return answer;
}

File: 서예진B/20230517/가장 큰 수.py
def solution(numbers):
    answer = ''
    ar = []
    # num은 최대 4자리수
    for num in numbers:
        # 각 숫자를 4번 반복해서 앞에서부터 4글자 슬라이싱
        '''
        처음에는 맨앞자리 수를 4자리가 될 때까지 채우면 되는 줄
        근데 만약 ['30', '303'] 이라면 '3033', '3033'이 되고
        만들 수 있는 가장 큰 수는 30330임 (30303보다 큼)
        그래서 '303'이 '30'보다 먼저 와야하는데 '3033'으로 같으므로 원하는대로 정렬 불가능
        '''
        tmp = (str(num) * 4)[:4]
        l = len(str(num))
        # ar배열에 변형시킨 문자열과 원래 문자열의 길이 저장
        ar.append((tmp, l))
        
    # 내림차순으로 정렬
    ar.sort(reverse=True)
    
    for (tmp, l) in ar:
        # 원래 문자열의 길이만큼 answer에 더해줌
        answer += tmp[:l]
        
    # answer의 결과가 int일 때 0이라면 ('00', '000','0000')
    if int(answer) == 0:
        # '0' 리턴
        return '0'
    
    return answer